### PR TITLE
Strategy from paper

### PR DESCRIPTION
**Pull request by rank-and-file**
Basic strategy from the paper

### CODE REVIEW

File: src/main/java/com/treasure/hunt/strategy/hint/impl/HalfPlaneHint.java [hunk 1]
```diff
@@ -1,107 +1,217 @@
...
  */
-public class HalfPlaneHint extends AngleHint {
 
+public class HalfPlaneHint extends AngleHint {
+    static final double visual_extent = 1000;
+    private LineString halfPlaneLine = null;
     @Getter
-    private Direction direction;    // when the line indicated by halfplanePointOne and halfplanePointTwo is not horizontal,
+    private Direction direction;
+    // when the line indicated by anglePointLeft and anglePointRight is not horizontal,
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/hint/impl/HalfPlaneHint.java]: This should be java doc

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            strategy.C = GEOMETRY_FACTORY.createPoint(newRectangle[2]);
+            strategy.D = GEOMETRY_FACTORY.createPoint(newRectangle[3]);
+            strategy.lastHintWasBad = false;
+            return moveToCenterOfRectangle(strategy.A, strategy.B, strategy.C, strategy.D, move);
+        } catch (Exception ee) {
+            printRect(strategy, rect, lastBadHint, curHint);
+            throw ee;
+        }
+
+    }
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java]: This method is way to long split it up

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+     * At, Bt, Ct and Dt in this implementation equate to A, B, C and D in the paper, since the
+     * not by phi transformed variables of the current rectangle R are also stored with A, B, C and D.
+     * The t signals the transformed state of these variables.
+     * hintT equates to the hint (L1', x1')
+     *
+     * @param curHint
+     * @return The move to scan various areas so that A,B,C and D can be updated to a smaller rectangle (or the treasure
+     * is found)
+     */
+    static Movement lastHintBadSubroutine(StrategyFromPaper strategy, HalfPlaneHint curHint, HalfPlaneHint lastBadHint,
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java]: Why is everything static , maybe make it a singleton or well u can also leave it

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        Point B = s.B;
+        Point C = s.C;
+        Point D = s.D;
+        System.out.println("A= (" + A.getX() + ", " + A.getY() + ")");
+        System.out.println(" B= (" + B.getX() + ", " + B.getY() + ")");
+        System.out.println(" C= (" + C.getX() + ", " + C.getY() + ")");
+        System.out.println(" D= (" + D.getX() + ", " + D.getY() + ")");
+        for (int i = 0; i < rect.length; i++)
+            System.out.println("rect[" + i + "]= " + rect[i]);
+        System.out.println("lastBadHint p1= " + lastBadHint.getCenter() + "lastHint p2= " +
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java]: 1 Finger

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    // some things get printed when exception occurs
+    static private void printRect(StrategyFromPaper s, Coordinate[] rect, HalfPlaneHint lastBadHint, HalfPlaneHint curHint) {
+        Point A = s.A;
+        Point B = s.B;
+        Point C = s.C;
+        Point D = s.D;
+        System.out.println("A= (" + A.getX() + ", " + A.getY() + ")");
+        System.out.println(" B= (" + B.getX() + ", " + B.getY() + ")");
+        System.out.println(" C= (" + C.getX() + ", " + C.getY() + ")");
+        System.out.println(" D= (" + D.getX() + ", " + D.getY() + ")");
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java]: Ganz viele Finger

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+        for (int i = 0; i < rect.length; i++)
+            System.out.println("rect[" + i + "]= " + rect[i]);
+        System.out.println("lastBadHint p1= " + lastBadHint.getCenter() + "lastHint p2= " +
+                lastBadHint.getRight());
+        System.out.println("curHint p1= " + curHint.getCenter() + "curHint p2= " +
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java]: Noch mehr Fnger

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        double min_y = Double.MAX_VALUE;
+        for (int i = 0; i < 4; i++) {
+            double x = rect[i].getX();
+            double y = rect[i].getY();
+
+            max_x = Math.max(max_x, x);
+            max_y = Math.max(max_y, y);
+            min_x = Math.min(min_x, x);
+            min_y = Math.min(min_y, y);
+
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java]: newline too much

File: src/test/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaperTest.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        assertPoints(points, true_points);
+
+        hint = new HalfPlaneHint(new Coordinate(2.5, 3), new Coordinate(-2.5, -4), right);
+        points = strat.move(hint).getPoints();
+        true_points = new Point[]{createPoint(0, -0.5), createPoint(0.75, -0.5)};
+        assertPoints(points, true_points);
+
+        hint = new HalfPlaneHint(new Coordinate(-4, -0.5), new Coordinate(4, -0.5), up);
+        points = strat.move(hint).getPoints();
+        true_points = new Point[]{createPoint(0.75, -0.5), createPoint(0.75, 1.25)};
```
Comment (1-alex98) [inline src/test/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaperTest.java]: Stick to variable naming conventions

File: src/test/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaperTest.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        assertPoints(points, true_points);
+
+        hint = new HalfPlaneHint(new Coordinate(-2.5, 3), new Coordinate(4, -0.5), right);
+        points = strat.move(hint).getPoints();
+        for (int i = 0; i < points.size(); i++) {
+            System.out.print(points.get(i).getObject());
+        }
+        System.out.println();
```
Comment (1-alex98) [inline src/test/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaperTest.java]: Finger

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/RoutinesFromPaper.java [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+    static Movement rectangleScan(Point A, Point B, Point C, Point D, Movement move) {
+        return rectangleScan(A.getCoordinate(), B.getCoordinate(), C.getCoordinate(), D.getCoordinate(), move);
+    }
+
+    static Movement rectangleScan(Coordinate A, Coordinate B, Coordinate C, Coordinate D, Movement move) {
+        int k = (int) A.distance(B);
+        Point[] a = new Point[k + 1];
+        Point[] b = new Point[k + 1];
+
+        { //create a_i on line segment AB
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/RoutinesFromPaper.java]: Well whatever this is supposed to be probably extract it into a method

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/RoutinesFromPaper.java [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+            double yDist = C.getY() - D.getY();
+            //for (int i = 0; i <= k; i++) {
+            for (int i = 0; i <= k; i++) {
+                b[i] = JTSUtils.createPoint(D.getX() + xDist * ((double) i / (double) k), D.getY() + yDist *
+                        ((double) i / (double) k));
+            }
+        }
+
+        if (k % 2 == 1) //code like in paper
+        {
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/RoutinesFromPaper.java]: Tick the format boxes in the commit message or str+alt+l { should be same lie as loop

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+     * If the rectangle is not parallel to the x and y axis, an error is thrown.
+     *
+     * @param rect
+     * @return
+     */
+    static Coordinate[] arrangeRectangle(Coordinate[] rect) {
+        assertRectangle(rect);
+
+        Coordinate A = null, B = null, C = null, D = null;
+        double max_x = -Double.MAX_VALUE;
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java]: var naming conventions

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+     * rect[1] is the upper right point, rect[2] is the bottom right point and rect[3] is the bottom left point.
+     * If the rectangle is not parallel to the x and y axis, an error is thrown.
+     *
+     * @param rect
+     * @return
+     */
+    static Coordinate[] arrangeRectangle(Coordinate[] rect) {
+        assertRectangle(rect);
+
+        Coordinate A = null, B = null, C = null, D = null;
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java]: What the hack are A B C And D

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
+import static com.treasure.hunt.utils.JTSUtils.doubleEqual;
+
+/**
+ * @author bsen
+ */
+public class GeometricUtils {
+
+    static void assertRectangle(Coordinate[] rect) {
+        if (rect.length != 4)
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java]: Then lets call it reactangle

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+                curHint.getRight());
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java]: Abbreviations are bad for your mental health

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+                    lineBetweenClockwise(L2_apos, L1_doubleApos, h_apos_g_apos)
+            ) {
+                // rectangleScan(phi_reverse(k, (s, s', d', d))
+                move = rectangleScanPhiReverse(basicTrans, rect, s, s_apos, d_apos, d, move);
+                // rectangleScan(phi_reverse(k, (g, g', h', h))
+                // newRectangle := Agpm
+                newRectangle = phiOtherRectangleInverse(basicTrans, rect,
+                        new Coordinate[]{At, g, p, m});
+            }
+            if ((x2_apos == left &&
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java]: Maybe a more descriptive name also namingconventions

**Review by 1-alex98 — CHANGES_REQUESTED** [review]
At least remove al System.outs

**Comment by 1-alex98** [comment]
Still not all old comments addressed at least address System.out.println

**Comment by rank-and-file** [comment]
> Still not all old comments addressed at least address System.out.println

Yes, I'm going to change that

**Pull request closed without merging**